File: csharp/MapleStory.Test/TMS/Character/TestGetCharacterSkill.cs
using NUnit.Framework;
using MapleStory.OpenAPI.TMS;
using MapleStory.OpenAPI.Common;

namespace MapleStory.Test.TMS.Character
{
    [TestFixture]
    public class TestGetCharacterSkill
    {
        private static readonly string apiKey = EnvConfig.Get("API_KEY_TMS"); // Your API Key
        private static readonly MapleStoryAPI api = new MapleStoryAPI(apiKey);
        private static readonly string ocid = "ab918948538b1b79046df133fff52092";

        [Test, Description("success: GetCharacterSkill")]
        public async Task GetCharacterSkill()
        {
            var skillGrade = "6";
            var response = await api.GetCharacterSkill(ocid, skillGrade);
            Assert.IsNotNull(response);
            Console.WriteLine(response.ToJson());
        }

        [Test, Description("success: GetCharacterSkill with date")]
        public async Task GetCharacterSkill_With_Date()
        {
            var skillGrade = "6";
            var date = new DateTime(2025, 10, 15, 0, 0, 0);
            var response = await api.GetCharacterSkill(ocid, skillGrade, date);
            Assert.IsNotNull(response);
            Console.WriteLine(response.ToJson());
        }

        [Test, Description("fail: GetCharacterSkill with invalid ocid throws OPENAPI00003")]
        public void GetCharacterSkill_With_Invalid_Ocid_Throws_OPENAPI00003()
        {
            var invalidOcid = "invalid_ocid_123";
            var skillGrade = "6";
            var e = Assert.ThrowsAsync<MapleStoryAPIException>(async () => await api.GetCharacterSkill(invalidOcid, skillGrade));
            Assert.That(e.ErrorCode, Is.EqualTo(MapleStoryAPIErrorCode.OPENAPI00003));
            Console.WriteLine($"{e.ErrorCode} {e.Message}");
        }

        [Test, Description("fail: GetCharacterSkill with invalid date")]
        public void GetCharacterSkill_With_Invalid_Date()
        {
            var skillGrade = "6";
            var invalidDate = new DateTime(2025, 10, 14, 0, 0, 0);
            var e = Assert.ThrowsAsync<ArgumentException>(async () => await api.GetCharacterSkill(ocid, skillGrade, invalidDate));
            Assert.That(e.Message, Does.Contain("You can only retrieve data after 2025-10-15."));
            Console.WriteLine(e.Message);
        }
    }
}

File: java/src/test/java/tms/character/TestGetCharacterSkill.java
package tms.character;

import dev.spiralmoon.maplestory.api.common.MapleStoryApiErrorCode;
import dev.spiralmoon.maplestory.api.common.MapleStoryApiException;
import dev.spiralmoon.maplestory.api.tms.MapleStoryApi;
import dev.spiralmoon.maplestory.api.tms.dto.character.CharacterSkillDTO;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

@DisplayName("getCharacterSkill")
public class TestGetCharacterSkill {
    private static final String apiKey = System.getProperty("API_KEY_TMS"); // Your API Key
    private static final MapleStoryApi api = new MapleStoryApi(apiKey);
    private static final String ocid = "ab918948538b1b79046df133fff52092";

    @Test
    @DisplayName("success: getCharacterSkill with skill grade")
    void getCharacterSkill() {
        String skillGrade = "6";
        CharacterSkillDTO response = api.getCharacterSkill(ocid, skillGrade).join();
        assertThat(response).isNotNull();
        System.out.println(response.toString());
    }

    @Test
    @DisplayName("success: async getCharacterSkill")
    void getCharacterSkill_async() {
        String skillGrade = "6";
        api.getCharacterSkill(ocid, skillGrade).thenAcceptAsync(response -> {
            assertThat(response).isNotNull();
            System.out.println(response.toString());
        }).join();
    }

    @Test
    @DisplayName("success: getCharacterSkill with date")
    void getCharacterSkill_with_date() {
        String skillGrade = "6";
        LocalDateTime date = LocalDateTime.of(2025, 10, 15, 0, 0);
        CharacterSkillDTO response = api.getCharacterSkill(ocid, skillGrade, date).join();
        assertThat(response).isNotNull();
        System.out.println(response.toString());
    }

    @Test
    @DisplayName("fail: getCharacterSkill with invalid ocid throw OPENAPI00003")
    void getCharacterSkill_with_invalid_ocid() {
        String invalidOcid = "invalid_ocid_123";
        String skillGrade = "6";
        assertThatThrownBy(() -> api.getCharacterSkill(invalidOcid, skillGrade).join())
                .hasCauseInstanceOf(MapleStoryApiException.class)
                .satisfies(e -> {
                    MapleStoryApiException apiException = (MapleStoryApiException) e.getCause();
                    assertThat(apiException.getErrorCode()).isEqualTo(MapleStoryApiErrorCode.OPENAPI00003);
                    System.out.println(apiException.getErrorCode() + " " + apiException.getMessage());
                });
    }

    @Test
    @DisplayName("fail: getCharacterSkill with invalid date")
    void getCharacterSkill_with_invalid_date() {
        String skillGrade = "6";
        LocalDateTime invalidDate = LocalDateTime.of(2025, 10, 14, 0, 0);
        assertThatThrownBy(() -> api.getCharacterSkill(ocid, skillGrade, invalidDate).join())
                .hasCauseInstanceOf(IllegalArgumentException.class)
                .satisfies(e -> {
                    Throwable error = e.getCause();
                    assertThat(error.getMessage()).contains("You can only retrieve data after 2025-10-15.");
                    System.out.println(error.getMessage());
                });
    }
}
